pkg/sql/pgwire/pgwirebase/encoding.go

```diff
@@ -102,9 +109,17 @@ func (b *ReadBuffer) ReadUntypedMsg(rd io.Reader) (int, error) {
 	size := int(binary.BigEndian.Uint32(b.tmp[:]))
 	// size includes itself.
 	size -= 4
-	if size > maxMessageSize || size < 0 {
-		return nread, NewProtocolViolationErrorf("message size %d out of bounds (0..%d)",
-			size, maxMessageSize)
+	if size < 0 {
+		err = NewProtocolViolationErrorf("invalid message size %d", size)
+	} else if size > maxMessageSize {
+		err = pgerror.Newf(pgcode.ProgramLimitExceeded,
+			"message of size %d bytes exceeds maximum of %d bytes. "+
+				"Try sending smaller batches.", size, maxMessageSize)
+	}
+	if err != nil {
+		// Consume the desired number of bytes to regain stream synchronization.
+		io.CopyN(ioutil.Discard, rd, int64(size))
```

Comment:
If `size < 0` we should probably not call this?

How did this work before? Is there a test to ensure this synchronization? It's unclear to me why we need this since `io.ReadFull` was called above.

Reply:
`ReadFull` reads `n` bytes, where `n` is equal to the size of the buffer you pass in. That buffer is of size 4. The synchronization is tested in the test I added now.